pulp_container/app/redirects.py

```diff
@@ -123,6 +91,47 @@ def redirect_to_object_storage(self, artifact, return_media_type):
         )
         return redirect(content_url)
 
+    # BACKWARD COMPATIBILITY - TODO remove after fully migrating to artifactless manifests
+    def redirect_to_artifact(self, content_name, manifest, manifest_media_type):
+        """
+        Search for the passed manifest's artifact and issue a redirect.
+        """
+        try:
+            artifact = manifest._artifacts.get()
+        except ObjectDoesNotExist:
+            raise Http404(f"An artifact for '{content_name}' was not found")
+
+        return self.redirect_to_object_storage(artifact, manifest_media_type)
+
+    def issue_tag_redirect(self, tag):
+        """
+        Issue a redirect if an accepted media type requires it or return not found if manifest
+        version is not supported.
+        """
+        if tag.tagged_manifest.data:
+            return super().issue_tag_redirect(tag)
+
+        manifest_media_type = tag.tagged_manifest.media_type
+        if manifest_media_type == MEDIA_TYPE.MANIFEST_V1:
+            return self.redirect_to_artifact(
+                tag.name, tag.tagged_manifest, MEDIA_TYPE.MANIFEST_V1_SIGNED
+            )
+        elif manifest_media_type in get_accepted_media_types(self.request.headers):
+            return self.redirect_to_artifact(tag.name, tag.tagged_manifest, manifest_media_type)
+        else:
+            raise ManifestNotFound(reference=tag.name)
+
+    def issue_manifest_redirect(self, manifest):
+        """
+        Directly redirect to an associated manifest's artifact.
+        """
+        if manifest.data:
+            return super().issue_manifest_redirect(manifest)
```

Comment:
As I am thinking more about this, it would be better to return manifest.data directly from registry_api.py. The code in registry.py will only need to deal with manifests that still have artifacts in place. The complexity of registry.py would be much lower because we could get rid of "BACKWARD COMPATIBILITY" piece. What do you think?

Reply:
I think this is a good idea, although I suggest postponing these changes to come with the PR that removes backward compatibility, along some other optimization changes that will be possible once backward compatibility is not needed anymore.